docs/src/proposals/interest-bearing-tokens.md

```diff
@@ -0,0 +1,414 @@
+interacting with. See the Associated Token Account section for how to use this.
+
+There are also new read-only instructions, which write data to a provided account
+buffer:
+
+* `AmountToUiAmount`: convert the given share amount to a UI token amount
+* `UiAmountToAmount`: convert the given UI token amount to an exact share amount
+
+See the Runtime section for more information on how these are used.
+
+Programs and wallets that wish to support i-tokens must update the instructions
+to use all `Checked` variants and use the new read-only instructions to fetch
+information.
+
+#### Conversions and Wrapper Library
+
+The SPL token library provides wrappers compatible with any conforming program.
+The wrapper simply checks the program id and deserializes the account data
+as an `spl_token::Holding` or `spl_token::Mint`. If a client program requires access
+to data specific to an i-token program, it needs to use that program's serializers.
+
+To convert between amount and UI amount, the wrapper library calls the `AmountToUiAmount`
+instruction on the i-token program, or for SPL token accounts, it calls
+`spl_token::amount_to_ui_amount` inline.
+
+For interest-bearing tokens, after deserialization into a `spl_token::Holding`,
+the `amount` field will be in terms of shares, and not token amount, so UIs and
+on-chain programs that need token amounts must convert using `AmountToUiAmount`.
+
+#### Struct Conformance
+
+The structure for new holding and mint accounts must follow the layout of the
+existing SPL token accounts to ensure compatibility in RPC when fetching pre and
+post token balances.
+
+For a holding, the first 165 bytes must contain the same information
+as a normal SPL token holding. The following byte must be the type of
+account (ie. `Holding`), and after that, any data is allowed as required by the
+new token program.
+
+The same applies for mints, but only for the first 82 bytes.
+
+### Token Program Registry
+
+We need the [token-list](https://github.com/solana-labs/token-list) to include
+vetted SPL token-conforming programs and group known mints by their program id.
+
+The token list will publish a file of known program ids at
+[token-list releases](https://github.com/solana-labs/token-list/releases/latest),
+to be used by RPC and programs.
+
+The entries in the token-list are signed using a managed key. As wallets or
+Ledgers are asked to send transactions using a particular token program, they can
+cross-check the mint or program information using a signature check on the 
+token-list data.
+
+The token-list client libraries can fetch, cache, and query all registry info,
+to be used for simple program and mint validation.
+
+TODO consider detailing the process of including a new program
+
+TODO figure out how the token-list signing key will be managed
+
+### Runtime
+
+#### Ephemeral Accounts
+
+In order to reduce friction when using read-only instructions on other programs,
```

Comment:
I think that having "set return value" and "get return value" syscalls would be a better approach and a better UX for devs designing programs which return computed information. Tag @jackcmay

Reply:
oh, yeah. I had same idea for awhile. :) cc: @joncinque

Reply:
I really like this idea and it would be very useful outside of interest bearing tokens. For example, it would allow CPI in Anchor to look like a normal function call.

Reply:
@seanyoung, @jackcmay, dmakarov  and I had a chat about this earlier today.  @seanyoung is going to write up a brief design proposal PR soon.  Basically we need the same for solang as well to return arbitrary data to the CPI caller